syft/pkg/cataloger/golang/version.go

```diff
@@ -1,226 +1,99 @@
-)
-
-func runVersion(ctx context.Context, cmd *base.Command, args []string) {
-	if len(args) == 0 {
-		// If any of this command's flags were passed explicitly, error
-		// out, because they only make sense with arguments.
-		//
-		// Don't error if the flags came from GOFLAGS, since that can be
-		// a reasonable use case. For example, imagine GOFLAGS=-v to
-		// turn "verbose mode" on for all Go commands, which should not
-		// break "go version".
-		var argOnlyFlag string
-		if !base.InGOFLAGS("-m") && *versionM {
-			argOnlyFlag = "-m"
-		} else if !base.InGOFLAGS("-v") && *versionV {
-			argOnlyFlag = "-v"
-		}
-		if argOnlyFlag != "" {
-			fmt.Fprintf(os.Stderr, "go: 'go version' only accepts %s flag with arguments\n", argOnlyFlag)
-			base.SetExitStatus(2)
-			return
-		}
-		fmt.Printf("go version %s %s/%s\n", runtime.Version(), runtime.GOOS, runtime.GOARCH)
-		return
-	}
-
-	for _, arg := range args {
-		info, err := os.Stat(arg)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "%v\n", err)
-			base.SetExitStatus(1)
-			continue
-		}
-		if info.IsDir() {
-			scanDir(arg)
-		} else {
-			scanFile(arg, info, true)
-		}
-	}
-}
-
-// scanDir scans a directory for executables to run scanFile on.
-func scanDir(dir string) {
-	filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
-		if d.Type().IsRegular() || d.Type()&fs.ModeSymlink != 0 {
-			info, err := d.Info()
-			if err != nil {
-				if *versionV {
-					fmt.Fprintf(os.Stderr, "%s: %v\n", path, err)
-				}
-				return nil
-			}
-			scanFile(path, info, *versionV)
-		}
-		return nil
-	})
+// unionReader is a single interface with all reading functions used by golang bin
+// cataloger.
+type unionReader interface {
+	io.Reader
+	io.ReaderAt
+	io.Seeker
+	io.Closer
 }
 
 // isExe reports whether the file should be considered executable.
-func isExe(file string, info fs.FileInfo) bool {
+func isExe(file string, mode os.FileMode) bool {
 	if runtime.GOOS == "windows" {
 		return strings.HasSuffix(strings.ToLower(file), ".exe")
 	}
-	return info.Mode().IsRegular() && info.Mode()&0111 != 0
+	return mode.IsRegular() && mode&0111 != 0
 }
 
 // scanFile scans file to try to report the Go and module versions.
-// If mustPrint is true, scanFile will report any error reading file.
-// Otherwise (mustPrint is false, because scanFile is being called
-// by scanDir) scanFile prints nothing for non-Go executables.
-func scanFile(file string, info fs.FileInfo, mustPrint bool) {
-	if info.Mode()&fs.ModeSymlink != 0 {
-		// Accept file symlinks only.
-		i, err := os.Stat(file)
-		if err != nil || !i.Mode().IsRegular() {
-			if mustPrint {
-				fmt.Fprintf(os.Stderr, "%s: symlink\n", file)
-			}
-			return
-		}
-		info = i
-	}
-
-	if !isExe(file, info) {
-		if mustPrint {
-			fmt.Fprintf(os.Stderr, "%s: not executable file\n", file)
-		}
-		return
+func scanFile(reader unionReader, filename string, mode os.FileMode) []*debug.BuildInfo {
+	if !isExe(filename, mode) {
```

Comment:
I think querying the cataloger already accounts for this kind of filtering (only requesting executables, though it does it by asking the resolve by MIME type instead of checking file extensions and mode values). So it shouldn't be needed. Was there a specific reason why an exe check was added (or kept around from previous code)?

Reply:
This func came from golang's file scan function, removing it